hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java

```diff
@@ -86,20 +125,63 @@ public HoodieMetadataPayload(Option<GenericRecord> record) {
       // https://issues.apache.org/jira/browse/AVRO-1811
       key = record.get().get(SCHEMA_FIELD_ID_KEY).toString();
       type = (int) record.get().get(SCHEMA_FIELD_ID_TYPE);
-      if (record.get().get(SCHEMA_FIELD_ID_METADATA) != null) {
-        filesystemMetadata = (Map<String, HoodieMetadataFileInfo>) record.get().get("filesystemMetadata");
+      if (record.get().get(SCHEMA_FIELD_ID_FILESYSTEM) != null) {
+        filesystemMetadata = (Map<String, HoodieMetadataFileInfo>) record.get().get(SCHEMA_FIELD_ID_FILESYSTEM);
         filesystemMetadata.keySet().forEach(k -> {
           GenericRecord v = filesystemMetadata.get(k);
           filesystemMetadata.put(k.toString(), new HoodieMetadataFileInfo((Long) v.get("size"), (Boolean) v.get("isDeleted")));
         });
       }
+
+      if (type == METADATA_TYPE_BLOOM_FILTER) {
```

Comment:
we need to pull the merging for each partition type into its own payload class and keep this just as a delegator. lets do this in this PR

Reply:
Based on the discussion, taking this up in https://issues.apache.org/jira/browse/HUDI-3323